### PR TITLE
[SB-12] Create and read favorites.

### PR DESCRIPTION
**Pull request by imariel2d**
**Things i did:**
- Refactored Favorites/Heart components.
- Fixed some styles.
- Fetch favorite songs.
- Create favorite songs.
- Delete favorite.

**Evidence**
https://www.loom.com/share/bee2e25ae2be405dabd3ea2713cba498

**Jira**
https://songbox-io.atlassian.net/browse/SB-12?atlOrigin=eyJpIjoiYWNlN2RmNzU1OTIxNGI1ZmEyYmE1NzViMjdlODIyZDkiLCJwIjoiaiJ9

### CODE REVIEW

File: src/components/Favorites/Favorite.component.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import { List } from 'immutable';
+import propTypes from 'prop-types';
+
+import LinkToSongContainer from '../Files/LinkToSong';
+
+const FavoriteComponent = ({ favorites }) => (
+  <div className="content-container">
+    <h1 id="your-personal-library" className="title">
```
Comment (tonymtz) [inline src/components/Favorites/Favorite.component.js]: Curious about the reason behind the existence of this id. I don't see where it is used.

File: src/components/Favorites/Favorite.component.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      Your personal library
+      <span role="img" aria-label="heart">❤️</span>
```
Comment (tonymtz) [inline src/components/Favorites/Favorite.component.js:13]: ~aria-label value does not make sense~

File: src/components/HeartFavorite/HeartFavorite.component.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import propTypes from 'prop-types';
+
+import './styles/favorite.scss';
+
+const HeartFavoriteComponent = ({ isFavorite, handleFavorite }) => {
```
Comment (tonymtz) [inline src/components/HeartFavorite/HeartFavorite.component.js]: nit: `onButtonClick`

File: src/components/HeartFavorite/HeartFavorite.container.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  const favorites = useSelector((state) => state.favorites.get('songs'));
+  const dispatch = useDispatch();
+
+  const handleFavorite = () => {
+    dispatch(createFavorite(fileName, path));
+  };
+
+  useEffect(() => {
+    setIsFavorite(isSongInFavorites(favorites, path));
```
Comment (tonymtz) [inline src/components/HeartFavorite/HeartFavorite.container.js]: I don't think you need a state for this. Also, if the song is not in favorites you are forcing a useless state update.

File: src/store/actions/favoritesActions.js [hunk 5]
```diff
@@ -1,3 +1,13 @@
+import { Map, fromJS } from 'immutable';
+import swal from 'sweetalert';
```
Comment (tonymtz) [inline src/store/actions/favoritesActions.js]: Let's avoid the use of this. We need to remove this dependency.

File: src/store/actions/favoritesActions.js [hunk 6]
```diff
@@ -8,6 +18,46 @@ import {
...
 export const initializeFavorites = (payload) => ({ type: FAVORITES_INITIALIZE_FAVORITES, payload });
+
+export const createFavorite = (fileName, path) => (dispatch, getState) => {
+  const token = getState().app.get('token');
+  const favorites = getState().favorites.get('songs');
+
+  const body = {
+    song_name: fileName,
+    path_lower: path,
+  };
```
Comment (tonymtz) [inline src/store/actions/favoritesActions.js]: Defer the body's creation to the service, not to the consumer.

File: src/store/actions/favoritesActions.js [hunk 7]
```diff
@@ -8,6 +18,46 @@ import {
...
+
+  if (!isSongInFavorites(favorites, path)) {
+    createFavoriteService(token, body)
+      .then(({ data }) => {
+        const songData = Map(data);
+        dispatch(addFavorite(songData));
+      })
+      .catch(() => {
```
Comment (tonymtz) [inline src/store/actions/favoritesActions.js]: All catches should report the error to rollbar (already implemented in the error logger).

File: src/utils/favorites.js [hunk 8]
```diff
@@ -0,0 +1 @@
+export const isSongInFavorites = (favorites, path) => favorites.some((favorite) => favorite.get('path_lower') === path);
```
Comment (tonymtz) [inline src/utils/favorites.js]: This belongs to the favorites module.

**Review by tonymtz — CHANGES_REQUESTED** [review]
I like the overall approach you followed, but I'm concerned you are repeating the same mistakes from past PRs.

File: src/components/Favorites/Favorite.component.js (hunk 1 above)
Comment (arielplas) [reply src/components/Favorites/Favorite.component.js]: Since we only have the route title in screen once, i thought it was a good idea to target its styles with id. It has a className 'title', yes. But that className is used to target font-family and font-weight. 😄

File: src/components/Favorites/Favorite.component.js (hunk 2 above)
Comment (arielplas) [reply src/components/Favorites/Favorite.component.js:13]: Based on this documentation (https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/accessible-emoji.md) it seems this is the correct way of naming it. If not, what do you suggest naming it? 🤔

Comment (tonymtz) [reply src/components/Favorites/Favorite.component.js:13]: Nvm, for some reason I misread this. I thought this was a button or something alike (?)

File: src/components/Favorites/Favorite.component.js (hunk 1 above)
Comment (tonymtz) [reply src/components/Favorites/Favorite.component.js]: Let's avoid using ids for styling purposes. Having ids will make CSS rules more complicated. Also, we are removing the current css implementation anyway.

File: src/components/Popup/Popup.component.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import propTypes from 'prop-types';
+
+import './styles/popup.scss';
+
+const PopupComponent = ({
```
Comment (tonymtz) [inline src/components/Popup/Popup.component.js]: Why did you create a popup component? 👀 

We do have material-ui and bootstrap. You literally have options there, and you chose to reinvent the wheel hahaha. Please revert this change, we don't need a popup.

**Review by tonymtz — CHANGES_REQUESTED** [review]
Please revert the last commit, we don't need a popup.

**Comment by tonymtz** [comment]
Please fix conflicts, also there's a couple of non-addressed comments @ImAriel2D

File: src/components/Favorites/Favorites.style.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import styled from 'styled-components';
+
+export const FavoritesContent = styled.div`
```
Comment (tonymtz) [inline src/components/Favorites/Favorites.style.js]: `FavoritesWrapper`

File: src/components/HeartFavorite/HeartFavorite.component.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
```
Comment (tonymtz) [inline src/components/HeartFavorite/HeartFavorite.component.js]: Check how icons are set here: https://github.com/tepachelabs/songbox.webapp/pull/33/files#diff-0c98ae846874305539a312f8b3606023eb76035f89dda93a48b9f0e6e6a169b0R1

Please try to follow the same pattern.

File: src/components/HeartFavorite/HeartFavorite.container.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import { useDispatch, useSelector } from 'react-redux';
+import propTypes from 'prop-types';
+
+import { handleInteractionWithFavorite } from 'store/actions/favoritesActions';
+
+import { isSongInFavorites } from 'Favorites/favorites';
+
+import HeartFavoriteComponent from './HeartFavorite.component';
+
+const HeartFavoriteContainer = ({ fileName, path }) => {
```
Comment (tonymtz) [inline src/components/HeartFavorite/HeartFavorite.container.js]: Let's move this logic to be part of the `LinkToSong` container

File: src/domains/core/app.container.js [hunk 13]
```diff
@@ -19,6 +21,8 @@ export const App = () => {
     } else {
       dispatch(recoverSession());
     }
+
+    dispatch(fetchFavoritesSongs());
```
Comment (tonymtz) [inline src/domains/core/app.container.js]: Dispatching this action here will cause `401` error (and an endless loop of redirects) if the user tries to access the `/login` route with not auth.

File: src/lib/apiService.js [hunk 14]
```diff
@@ -47,6 +47,20 @@ export const postWithAuth = (url, token, body, resolve, reject) => {
     .catch((error) => onError(error, reject));
 };
 
+export const deleteWithAuth = (url, token, body, resolve, reject) => {
```
Comment (tonymtz) [inline src/lib/apiService.js:50]: Nice addition.

File: src/components/Favorites/Favorites.style.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+ width: 100%;
+`;
+
+export const FavoritesSongs = styled.div`
```
Comment (tonymtz) [inline src/components/Favorites/Favorites.style.js]: `FavoritesSongList` or `FavoritesSongListWrapper`

**Review by tonymtz — CHANGES_REQUESTED** [review]
I liked how you solved the communication with the API.

I'm blocking because we need to update the icon and to move the place from where favorites are initially fetched.

File: src/components/Favorites/Favorites.style.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import styled from 'styled-components';
+
+export const FavoritesWrapper = styled.div`
+ width: 100%;
+`;
+
+export const FavoritesSongListWrapper = styled.div`
+  padding: 0 20px 150px 0;
```
Comment (tonymtz) [inline src/components/Favorites/Favorites.style.js]: Let's use the theme's spacing instead of arbitrary numbers.

File: src/components/Files/LinkToSong/LinkToSong.container.js [hunk 17]
```diff
@@ -8,6 +8,8 @@ import { getSongStreamLink } from 'store/actions/playerActions';
 
 import { selectSongPathAtIndex } from 'store/selectors/songsQueue';
 import LinkToSongComponent from './LinkToSong.component';
+import { isSongInFavorites } from '../../../Favorites/favorites';
+import { handleInteractionWithFavorite } from '../../../store/actions/favoritesActions';
```
Comment (tonymtz) [inline src/components/Files/LinkToSong/LinkToSong.container.js]: use relative imports instead

File: src/components/HeartFavorite/HeartFavorite.style.js [hunk 18]
```diff
@@ -0,0 +1,10 @@
+import styled from 'styled-components';
+
+export const HeartButton = styled.button`
+  background: none;
+  border: none;
+`;
```
Comment (tonymtz) [inline src/components/HeartFavorite/HeartFavorite.style.js]: You should create a new Button variant instead, check how `DropboxButton` is created.

File: src/components/Favorites/Favorite.component.js [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+import {
+  FavoritesWrapper,
+  FavoritesTitle,
+  FavoritesSongListWrapper,
+} from './Favorites.style';
+
+import LinkToSongContainer from '../Files/LinkToSong';
+
+const FavoriteComponent = ({ favorites }) => (
```
Comment (tonymtz) [inline src/components/Favorites/Favorite.component.js:9]: Let's use Grid elements as possible instead of custom wrappers.

https://material-ui.com/components/grid/

File: src/components/HeartFavorite/HeartFavorite.style.js (hunk 18 above)
Comment (jmsalcido) [reply src/components/HeartFavorite/HeartFavorite.style.js]: thats the entire point of the styled components @ImAriel2D

**Review by jmsalcido — CHANGES_REQUESTED** [review]

File: src/components/Favorites/Favorites.style.js [hunk 20]
```diff
@@ -0,0 +1,8 @@
+import styled from 'styled-components';
+import { theme } from 'style/theme';
+
+export const FavoritesTitle = styled.h1`
+  font-size: ${theme.spacing.three};
+  text-align: left;
+  padding: ${theme.spacing.two};
```
Comment (tonymtz) [inline src/components/Favorites/Favorites.style.js]: Do **NOT** import the theme, it's already available as prop.

File: src/components/Files/LinkToSong/LinkToSong.component.js [hunk 21]
```diff
@@ -21,8 +22,8 @@ const LinkToSongComponent = ({
       </button>
       <Fragment>
         <HeartFavorite
-          fileName={fileName}
-          path={path}
+          isFavorite={isFavorite}
+          onButtonClick={handleFavorite}
```
Comment (tonymtz) [inline src/components/Files/LinkToSong/LinkToSong.component.js]: `onClick`

File: src/components/Files/LinkToSong/LinkToSong.container.js [hunk 22]
```diff
@@ -17,23 +19,32 @@ const LinkToSongContainer = ({
...
+  const isPlaying = path === playingSongPath;
+
   const selectSong = () => {
     dispatch(getSongStreamLink(path));
     dispatch(setSongIndex(index));
     dispatch(changeSongsQueue(files));
   };
 
-  const isPlaying = path === playingSongPath;
+  const handleFavorite = () => {
```
Comment (tonymtz) [inline src/components/Files/LinkToSong/LinkToSong.container.js]: `onAddFavorite`

File: src/components/Files/LinkToSong/LinkToSong.container.js [hunk 23]
```diff
@@ -17,23 +19,32 @@ const LinkToSongContainer = ({
...
+    dispatch(handleInteractionWithFavorite(isFavorite, {
+      song_name: fileName,
+      path_lower: path,
```
Comment (tonymtz) [inline src/components/Files/LinkToSong/LinkToSong.container.js]: use camel case for variables, rename later as part of the marshal/unmarshal routine.

File: src/style/button.js [hunk 24]
```diff
@@ -19,3 +19,10 @@ export const DropboxButton = withStyles(() => ({
     },
   },
 }))(Button);
+
+export const HeartButton = withStyles(() => ({
```
Comment (tonymtz) [inline src/style/button.js]: Please use a less specific name for this new variant. Let's aim at reusability.

File: src/domains/core/workspace/workspace.container.js [hunk 25]
```diff
@@ -1,13 +1,19 @@
...
 import { WorkspaceComponent } from './workspace.component';
 import { selectIsSessionToken } from '../app.selectors';
 
 export const Workspace = () => {
   const isSessionToken = useSelector(selectIsSessionToken);
+  const dispatch = useDispatch();
+
+  if (isSessionToken) {
+    dispatch(fetchFavoritesSongs());
+  }
```
Comment (tonymtz) [inline src/domains/core/workspace/workspace.container.js]: This is not good. Please take a look at this:

https://reactjs.org/docs/hooks-reference.html

File: src/style/colors.js [hunk 26]
```diff
@@ -6,6 +6,7 @@ export const orange = '#ffa834';
 export const lightGray = '#d2dae2';
 export const almostWhite = '#f0f4f7';
 export const darkWhite = '#b3b3ba';
+export const softGray = '#8A8A8A';
```
Comment (tonymtz) [inline src/style/colors.js]: I don't remember this color. Please do not add new colors that don't belong to the mocks.

File: src/lib/apiFavoritesService.js [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+  getWithAuth,
+  deleteWithAuth,
+} from 'lib/apiService';
+
+import {
+  API_FAVORITE_ROUTE,
+  API_FAVORITES_ROUTE,
+} from 'routes';
+
+export const createFavoriteService = (token, body) => new Promise((resolve, reject) => {
```
Comment (tonymtz) [inline src/lib/apiFavoritesService.js]: `apiAddFavorite`

The verb "create" is not wrong at all. However, this is not a service.

File: src/lib/apiFavoritesService.js [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+  postWithAuth(API_FAVORITE_ROUTE, token, body, resolve, reject);
+});
+
+export const getFavoriteSongsService = (token) => new Promise((resolve, reject) => {
```
Comment (tonymtz) [inline src/lib/apiFavoritesService.js]: `apiFetchFavorites`

File: src/lib/apiFavoritesService.js [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+  getWithAuth(API_FAVORITES_ROUTE, token, resolve, reject);
+});
+
+export const deleteFavoriteSongService = (token, body) => new Promise((resolve, reject) => {
```
Comment (tonymtz) [inline src/lib/apiFavoritesService.js]: `apiDeleteFavorite`

**Review by tonymtz — CHANGES_REQUESTED** [review]
A few comments added.

File: src/style/button.js (hunk 24 above)
Comment (arielplas) [reply src/style/button.js]: What name do you suggest? 😄

Comment (tonymtz) [reply src/style/button.js]: `BordelessButton` or `TransparentButton`.

It's a great addition, I think we will use this a lot 👍🏼

File: src/utils/transforms.js [hunk 30]
```diff
@@ -0,0 +1,11 @@
+/*
+* TODO: Fix this temporary fix
+*
+* This functions is a temporary fix
+* server receives `song_name` and `path_lower`
+*
+*/
+export const serializeFavoriteObject = (body) => ({
```
Comment (tonymtz) [inline src/utils/transforms.js:8]: awesome 👍🏼

File: src/domains/core/workspace/workspace.container.js [hunk 31]
```diff
@@ -1,13 +1,19 @@
...
 import { WorkspaceComponent } from './workspace.component';
 import { selectIsSessionToken } from '../app.selectors';
 
 export const Workspace = () => {
   const isSessionToken = useSelector(selectIsSessionToken);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(fetchFavoritesSongs());
+  }, [dispatch]);
```
Comment (tonymtz) [inline src/domains/core/workspace/workspace.container.js:16]: You nailed it.

**Review by tonymtz — APPROVED** [review]
### LGTM!!!!!!!!!!!!!

File: src/components/Favorites/Favorites.style.js [hunk 32]
```diff
@@ -0,0 +1,7 @@
+import styled from 'styled-components';
+
+export const FavoritesTitle = styled.h1`
```
Comment (jmsalcido) [inline src/components/Favorites/Favorites.style.js:3]: I am confused a little bit, why this lives below `../Favorites/` instead of having a style for audio player, Im not 100% sure of how we separate these styles now.

cc @tonymtz

**Review by jmsalcido — APPROVED** [review]
:shipit:, I've got a minor question about style file but can be talked over slack or in person, not a blocker.
